okapi-core/src/main/java/org/folio/okapi/managers/ProxyService.java

```diff
@@ -1313,13 +1304,37 @@ public void redirectProxy(RoutingContext ctx) {
         .replaceFirst("^/_/invoke/tenant/([^/ ]+)/.*$", "$1");
     String newPath = origPath
         .replaceFirst("^/_/invoke/tenant/[^/ ]+(/.*$)", "$1");
-    if (qry != null && !qry.isEmpty()) {
-      // vert.x 3.5 clears the parameters on reroute, so we pass them in ctx
-      ctx.data().put(REDIRECTQUERY, qry);
+
+    // delegate CORS for preflight request
+    if (HttpMethod.OPTIONS.equals(ctx.request().method())
+        && ctx.data().containsKey(CorsHelper.DELEGATE_CORS)) {
+      ModuleInstance mi = (ModuleInstance) ctx.data().get(CorsHelper.DELEGATE_CORS_MODULE_INSTANCE);
+      resolveUrls(Arrays.asList(mi));
+      RequestOptions requestOptions = new RequestOptions().setMethod(ctx.request().method())
+          .setAbsoluteURI(mi.getUrl() + newPath);
+      Future<HttpClientRequest> fut = httpClient.request(requestOptions);
+      fut.onFailure(cause -> HttpResponse.responseError(ctx, 500, cause.getMessage()));
+      fut.onSuccess(clientRequest -> {
+        copyHeaders(clientRequest, ctx, null);
+        clientRequest.end();
+        clientRequest.response()
+            .onFailure(cause -> HttpResponse.responseError(ctx, 500, cause.getMessage()))
+            .onSuccess(res -> {
+              ctx.response().setStatusCode(res.statusCode());
+              ctx.response().headers().addAll(res.headers());
+              sanitizeAuthHeaders(ctx.response().headers());
+              ctx.response().end();
+            });
+      });
```

Comment:
```suggestion
      httpClient.request(requestOptions)
          .compose(clientRequest -> {
            copyHeaders(clientRequest, ctx, null);
            clientRequest.end();
            return clientRequest.response();
          })
          .onSuccess(res -> {
            ctx.response().setStatusCode(res.statusCode());
            ctx.response().headers().addAll(res.headers());
            sanitizeAuthHeaders(ctx.response().headers());
            ctx.response().end();
          })
          .onFailure(cause -> HttpResponse.responseError(ctx, 500, cause.getMessage()));
```
Replacing onSuccess by compose avoids nesting and is easier to read and understand.
This also allows to drop the duplicate onFailure code.